include/dba.php

```diff
@@ -497,6 +441,272 @@ public function any_value_fallback($sql) {
 		}
 		return $sql;
 	}
+
+	/**
+	 * @brief Executes a prepared statement
+	 *
+	 * @param string $sql SQL statement
+	 * @return object statement object
+	 */
+	static public function p($sql) {
+		$a = get_app();
+
+		$stamp1 = microtime(true);
+
+		$args = func_get_args();
+		unset($args[0]);
+
+		if (!self::$dbo OR !self::$dbo->connected) {
+			return false;
+		}
+
+		$sql = self::$dbo->any_value_fallback($sql);
+
+		if (x($a->config,'system') && x($a->config['system'], 'db_callstack')) {
+			$sql = "/*".$a->callstack()." */ ".$sql;
+		}
+
+		switch (self::$dbo->driver) {
+			case 'pdo':
+				if (!$stmt = self::$dbo->db->prepare($sql)) {
+					$errorInfo = self::$dbo->db->errorInfo();
+					self::$dbo->error = $errorInfo[2];
+					self::$dbo->errorno = $errorInfo[1];
+					$retval = false;
+					break;
+				}
+
+				foreach ($args AS $param => $value) {
+					$stmt->bindParam($param, $args[$param]);
+				}
+
+				if (!$stmt->execute()) {
+					$errorInfo = self::$dbo->db->errorInfo();
+					self::$dbo->error = $errorInfo[2];
+					self::$dbo->errorno = $errorInfo[1];
+					$retval = false;
+				} else {
+					$retval = $stmt;
+				}
+				break;
+			case 'mysqli':
+				$stmt = self::$dbo->db->stmt_init();
+
+				if (!$stmt->prepare($sql)) {
+					self::$dbo->error = self::$dbo->db->error;
+					self::$dbo->errorno = self::$dbo->db->errno;
+					$retval = false;
+					break;
+				}
+
+				$params = '';
+				$values = array();
+				foreach ($args AS $param => $value) {
+					if (is_int($args[$param])) {
+						$params .= 'i';
+					} elseif (is_float($args[$param])) {
+						$params .= 'd';
+					} elseif (is_string($args[$param])) {
+						$params .= 's';
+					} else {
+						$params .= 'b';
+					}
+					$values[] = &$args[$param];
+				}
+
+				array_unshift($values, $params);
+
+				call_user_func_array(array($stmt, 'bind_param'), $values);
+
+				if (!$stmt->execute()) {
+					self::$dbo->error = self::$dbo->db->error;
+					self::$dbo->errorno = self::$dbo->db->errno;
+					$retval = false;
+				} else {
+					$stmt->store_result();
+					$retval = $stmt;
+				}
+				break;
+			case 'mysql':
+				// For the old "mysql" functions we cannot use prepared statements
+				foreach ($args AS $param => $value) {
+					if (is_int($args[$param]) OR is_float($args[$param])) {
+						$replace = intval($args[$param]);
+					} else {
+						$replace = "'".dbesc($args[$param])."'";
+					}
+
+					$pos = strpos($sql, '?');
+					if ($pos !== false) {
+						$sql = substr_replace($sql, $replace, $pos, 1);
```

Comment:
I understand the goal, but this is broken if any of the replaced values holds a question mark. Ideally, you should advance `$pos` to the end of the replaced string, and use it as a starting point for the next `strpos()`.